File: platform/platform-impl/src/com/intellij/ide/ui/laf/PluggableLafInfo.java
// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.ide.ui.laf;

import com.intellij.ide.ui.laf.darcula.ui.DarculaEditorTextFieldBorder;
import com.intellij.openapi.editor.ex.EditorEx;
import com.intellij.ui.EditorTextField;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import javax.swing.text.JTextComponent;

public abstract class PluggableLafInfo extends UIManager.LookAndFeelInfo {
  public PluggableLafInfo(String name, String className) {
    super(name, className);
  }

  public static class SearchAreaContext {
    private final JComponent searchTextArea;
    private final JTextComponent textComponent;
    private final JComponent iconsPanel;
    private final JComponent scrollPane;

    public SearchAreaContext(JComponent searchTextArea, JTextComponent textComponent,
                             JComponent iconsPanel, JComponent scrollPane) {
      this.searchTextArea = searchTextArea;
      this.textComponent = textComponent;
      this.iconsPanel = iconsPanel;
      this.scrollPane = scrollPane;
    }

    public JComponent getSearchComponent() {
      return searchTextArea;
    }

    public JTextComponent getTextComponent() {
      return textComponent;
    }

    public JComponent getIconsPanel() {
      return iconsPanel;
    }

    public JComponent getScrollPane() {
      return scrollPane;
    }
  }

  public abstract SearchTextAreaPainter createSearchAreaPainter(@NotNull SearchAreaContext context);
  public abstract DarculaEditorTextFieldBorder createEditorTextFieldBorder(EditorTextField editorTextField, EditorEx editor);
}

File: platform/platform-impl/src/com/intellij/ide/ui/laf/PluggableLafInfo.kt
// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.ide.ui.laf

import com.intellij.ide.ui.laf.darcula.ui.DarculaEditorTextFieldBorder
import com.intellij.openapi.editor.ex.EditorEx
import com.intellij.ui.EditorTextField
import javax.swing.JComponent
import javax.swing.LookAndFeel
import javax.swing.UIDefaults
import javax.swing.UIManager
import javax.swing.text.JTextComponent

abstract class PluggableLafInfo(name: String, className: String) : UIManager.LookAndFeelInfo(name, className) {
  class SearchAreaContext(val searchComponent: JComponent, val textComponent: JTextComponent,
                          val iconsPanel: JComponent, val scrollPane: JComponent)
  /**
   * Creates an instance of the LookAndFeel provided by the plugin.
   */
  abstract fun createLookAndFeel() : LookAndFeel

  /**
   * Returns <code>true</code> if the LookAndFeel is dark, <code>false</code> otherwise.
   */
  open fun isDark() : Boolean = false

  /**
   * Creates paint context needed to render `SearchTextArea` with this look and feel.
   * @param context is wrapper object that contains all fields that may potentially be needed for
   * rendering.
   * @return a new instance of the painter.
   */
  abstract fun createSearchAreaPainter(context: SearchAreaContext): SearchTextAreaPainter

  /**
   * Creates an instance of `Border` that paints decorations and focus border
   * around EditorTextField
   * @param editorTextField
   * @param editor
   * @return a new instance of the border.
   */
  abstract fun createEditorTextFieldBorder(editorTextField: EditorTextField, editor: EditorEx): DarculaEditorTextFieldBorder
}